kv/kvserver: TestProcessSplitAfterRightHandSideHasBeenRemoved failed

kv/kvserver.TestProcessSplitAfterRightHandSideHasBeenRemoved [failed](https://mesolite.cluster.engflow.com/invocations/default/a8746d11-a8b4-4d08-ae3c-df6471ece88c?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [299240d06a0bf6036c92ca18da148255342fa516](https://github.com/cockroachdb/cockroach/commits/299240d06a0bf6036c92ca18da148255342fa516):


```
=== RUN   TestProcessSplitAfterRightHandSideHasBeenRemoved
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestProcessSplitAfterRightHandSideHasBeenRemoved3936177541
    test_log_scope.go:82: use -show-logs to present logs inline
    client_raft_test.go:5759: -- test log scope end --
test logs left over in: outputs.zip/logTestProcessSplitAfterRightHandSideHasBeenRemoved3936177541
--- FAIL: TestProcessSplitAfterRightHandSideHasBeenRemoved (412.44s)
=== RUN   TestProcessSplitAfterRightHandSideHasBeenRemoved/lease-type=LeaseExpiration
--- FAIL: TestProcessSplitAfterRightHandSideHasBeenRemoved/lease-type=LeaseExpiration (291.09s)
=== RUN   TestProcessSplitAfterRightHandSideHasBeenRemoved/lease-type=LeaseExpiration/(4)_initial_replica_RHS_partition,_with_restart
    client_raft_test.go:5741: condition failed to evaluate within 3m45s: from client_raft_test.go:5744: hard state not yet initialized: got 0, expected 10
--- FAIL: TestProcessSplitAfterRightHandSideHasBeenRemoved/lease-type=LeaseExpiration/(4)_initial_replica_RHS_partition,_with_restart (246.45s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestProcessSplitAfterRightHandSideHasBeenRemoved.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54862

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -5396,6 +5396,11 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 							// Newly-started stores (including the "rogue" one) should not GC
 							// their replicas. We'll turn this back on when needed.
 							DisableReplicaGCQueue: true,
+							// Some subtests, e.g. (4), expect that n1 catches up on the raft
+							// log after a partition, and runs the split trigger. Disable raft
+							// log truncation to make sure that it doesn't miss the split
+							// trigger by being caught up by a later snapshot. See #154313.
+							DisableRaftLogQueue:   true,
 							TestingProposalFilter: testingProposalFilter,
 						},
 					},

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -33,6 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/livenesspb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/rditer"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/stateloader"
 	raft "github.com/cockroachdb/cockroach/pkg/raft"
@@ -5320,13 +5321,13 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 			t *testing.T, store *kvserver.Store, rangeID roachpb.RangeID,
 		) raftpb.HardState {
 			t.Helper()
-			hs, err := stateloader.Make(rangeID).LoadHardState(ctx, store.TODOEngine())
+			hs, err := logstore.NewStateLoader(rangeID).LoadHardState(ctx, store.LogEngine())
 			require.NoError(t, err)
 			return hs
 		}
 		partitionReplicaOnSplit := func(t *testing.T, tc *testcluster.TestCluster, key roachpb.Key, basePartition *testClusterPartitionedRange, partRange **testClusterPartitionedRange) {
-			// Set up a hook to partition the RHS range at its initial range ID
-			// before proposing the split trigger.
+			// Set up a hook to partition away the first store of the RHS range at the
+			// first opportunity (when the split trigger is proposed).
 			var setupOnce sync.Once
 			f := kvserverbase.ReplicaProposalFilter(func(args kvserverbase.ProposalFilterArgs) *kvpb.Error {
 				req, ok := args.Req.GetArg(kvpb.EndTxn)
@@ -5357,9 +5358,10 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 			proposalFilter.Store(f)
 		}
 
-		// The basic setup for all of these tests are that we have a LHS range on 3
-		// nodes and we've partitioned store 0 for the LHS range. The tests will now
-		// perform a split, remove the RHS, add it back and validate assumptions.
+		// The basic setup for all of these tests are that we have an LHS range on 3
+		// nodes (lease on the last one), and we've partitioned store 0 for the LHS
+		// range. The tests will now perform a split, remove the RHS, add it back
+		// and validate assumptions.
 		//
 		// Different outcomes will occur depending on whether and how the RHS is
 		// partitioned and whether the server is killed. In all cases we want the
@@ -5480,9 +5482,9 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 			target := tc.Target(0)
 			log.KvExec.Infof(ctx, "removing voter: %v", target)
 			tc.RemoveVotersOrFatal(t, keyB, target)
-			// Unsuccessful because the RHS will not accept the learner snapshot
-			// and will be rolled back. Nevertheless it will have learned that it
-			// has been removed at the old replica ID.
+			// Unsuccessful because the RHS will not accept the learner snapshot and
+			// will be rolled back. Nevertheless, it will have learned that it has
+			// been removed at the old replica ID.
 			_, err = tc.Servers[0].DB().AdminChangeReplicas(
 				ctx, keyB, tc.LookupRangeOrFatal(t, keyB),
 				kvpb.MakeReplicationChanges(roachpb.ADD_VOTER, target),
@@ -5695,8 +5697,8 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 			target := tc.Target(0)
 			log.KvExec.Infof(ctx, "removing voter: %v", target)
 			tc.RemoveVotersOrFatal(t, keyB, target)
-			// Unsuccessfuly because the RHS will not accept the learner snapshot
-			// and will be rolled back. Nevertheless it will have learned that it
+			// Unsuccessful because the RHS will not accept the learner snapshot
+			// and will be rolled back. Nevertheless, it will have learned that it
 			// has been removed at the old replica ID.
 			//
 			// Not using tc.AddVoters because we expect an error, but that error
@@ -5716,7 +5718,7 @@ func TestProcessSplitAfterRightHandSideHasBeenRemoved(t *testing.T) {
 			log.KvExec.Infof(ctx, "added %d to RHS partition: %s", rhsInfo.Desc.NextReplicaID, rhsPartition)
 
 			// We do all of this incrementing to ensure that nobody will ever
-			// succeed in sending a message the new RHS replica after we restart
+			// succeed in sending a message to the new RHS replica after we restart
 			// the store. Previously there were races which could happen if we
 			// stopped the store immediately. Sleeps worked but this feels somehow
 			// more principled.

